pkg/module/go_get_fetcher.go

```diff
@@ -0,0 +1,159 @@
+	goPathRoot, err := afero.TempDir(g.fs, "", "athens")
+	if err != nil {
+		// TODO: return a ref for cleaning up the goPathRoot
+		return nil, err
+	}
+	sourcePath := filepath.Join(goPathRoot, "src")
+	modPath := filepath.Join(sourcePath, getRepoDirName(g.repoURI, g.version))
+	if err := g.fs.MkdirAll(modPath, os.ModeDir|os.ModePerm); err != nil {
+		// TODO: return a ref for cleaning up the goPathRoot
+		return nil, err
+	}
+
+	// setup the module with barebones stuff
+	if err := prepareStructure(g.fs, modPath); err != nil {
+		// TODO: return a ref for cleaning up the goPathRoot
+		return nil, err
+	}
+
+	cachePath, err := getSources(g.goBinaryName, g.fs, goPathRoot, modPath, mod, ver)
+	if err != nil {
+		// TODO: return a ref that cleans up the goPathRoot
+		return newDiskRef(g.fs, cachePath, ver), err
+	}
+	// TODO: make sure this ref also cleans up the goPathRoot
+	diskRef := newDiskRef(g.fs, cachePath, ver)
+
+	return diskRef, err
+}
+
+// Hacky thing makes vgo not to complain
+func prepareStructure(fs afero.Fs, repoRoot string) error {
+	// vgo expects go.mod file present with module statement or .go file with import comment
+	gomodPath := filepath.Join(repoRoot, "go.mod")
+	gomodContent := []byte("module mod")
+	if err := afero.WriteFile(fs, gomodPath, gomodContent, 0666); err != nil {
+		return err
+	}
+
+	sourcePath := filepath.Join(repoRoot, "mod.go")
+	sourceContent := []byte(`package mod // import "mod"`)
+	return afero.WriteFile(fs, sourcePath, sourceContent, 0666)
+}
+
+// given a filesystem, gopath, repository root, module and version, runs 'vgo get'
+// on module@version from the repoRoot with GOPATH=gopath, and returns the location
+// of the module cache. returns a non-nil error if anything went wrong. always returns
+// the location of the module cache so you can delete it if necessary
+func getSources(goBinaryName string, fs afero.Fs, gopath, repoRoot, module, version string) (string, error) {
+	version = strings.TrimPrefix(version, "@")
+	if !strings.HasPrefix(version, "v") {
+		version = "v" + version
+	}
+	uri := strings.TrimSuffix(module, "/")
+
+	fullURI := fmt.Sprintf("%s@%s", uri, version)
+
+	gopathEnv := fmt.Sprintf("GOPATH=%s", gopath)
+	cacheEnv := fmt.Sprintf("GOCACHE=%s", filepath.Join(gopath, "cache"))
+	disableCgo := "CGO_ENABLED=0"
+
+	cmd := exec.Command(goBinaryName, "get", fullURI)
+	// PATH is needed for vgo to recognize vcs binaries
+	// this breaks windows.
+	cmd.Env = []string{"PATH=" + os.Getenv("PATH"), gopathEnv, cacheEnv, disableCgo}
+	cmd.Dir = repoRoot
+
+	packagePath := filepath.Join(gopath, "src", "mod", "cache", "download", module, "@v")
+
+	o, err := cmd.CombinedOutput()
+	if err != nil {
+		// github quota exceeded
+		if isLimitHit(o) {
+			return packagePath, pkgerrors.WithMessage(err, "github API limit hit")
+		}
+		// another error in the output
+		return packagePath, err
+	}
+	// make sure the expected files exist
+	if err := checkFiles(fs, packagePath, version); err != nil {
```

Comment:
This could be simplified to just
```go
return packagePath, checkFiles(fs, packagePath, version)
```
as you always return `packagePath` and the error will be nil or not executing `checkFiles`.

Reply:
Here I was thinking the above can replace all of lines 131-134

Reply:
I replaced the `if err := checkFiles` and did what you recommended in https://github.com/gomods/athens/pull/291#discussion_r204963367. I felt like simplifying the `if err != nil` block above would hurt readability though. Cool with you?

Reply:
:+1: